packages/ui/src/hooks/useBreakpoint.ts

```diff
@@ -25,40 +25,111 @@ export const breakpoints: { name: string; id: Breakpoint; value: number }[] = [
-  );
-
-  const downMatches = new Map(
-    breakpoints.map(bp => [
-      bp.id,
-      useMediaQuery(`(max-width: ${bp.value - 1}px)`),
-    ]),
-  );
-
-  let breakpoint: Breakpoint = breakpoints[0].id;
-  for (let i = matches.length - 1; i >= 0; i--) {
-    if (matches[i]) {
-      breakpoint = breakpoints[i].id;
-      break;
+  useEffect(() => {
+    if (typeof window === 'undefined') return;
+
+    const mediaQuery = window.matchMedia(query);
+    const handleChange = () => setMatches(mediaQuery.matches);
+
+    // Set initial value
+    setMatches(mediaQuery.matches);
+
+    // Add listener
+    if (mediaQuery.addListener) {
+      mediaQuery.addListener(handleChange);
+    } else {
+      mediaQuery.addEventListener('change', handleChange);
     }
-  }
+
+    return () => {
+      if (mediaQuery.removeListener) {
+        mediaQuery.removeListener(handleChange);
+      } else {
+        mediaQuery.removeEventListener('change', handleChange);
+      }
+    };
+  }, [query]);
+
+  return matches;
+}
+
+// Hook to get current breakpoint efficiently - only triggers on breakpoint changes
+function useCurrentBreakpoint(): Breakpoint {
+  const [breakpoint, setBreakpoint] = useState<Breakpoint>(() => {
+    if (typeof window === 'undefined') return 'initial';
+
+    const width = window.innerWidth;
+    if (width >= 1536) return 'xl';
+    if (width >= 1280) return 'lg';
+    if (width >= 1024) return 'md';
+    if (width >= 768) return 'sm';
+    if (width >= 640) return 'xs';
+    return 'initial';
+  });
+
+  useEffect(() => {
+    if (typeof window === 'undefined') return;
+
+    const handleResize = () => {
+      const width = window.innerWidth;
+      let newBreakpoint: Breakpoint;
+
+      if (width >= 1536) newBreakpoint = 'xl';
+      else if (width >= 1280) newBreakpoint = 'lg';
+      else if (width >= 1024) newBreakpoint = 'md';
+      else if (width >= 768) newBreakpoint = 'sm';
+      else if (width >= 640) newBreakpoint = 'xs';
+      else newBreakpoint = 'initial';
+
+      setBreakpoint(prev => (prev !== newBreakpoint ? newBreakpoint : prev));
+    };
+
+    window.addEventListener('resize', handleResize);
+    return () => window.removeEventListener('resize', handleResize);
+  }, []);
+
+  return breakpoint;
+}
+
+/** @public */
+export const useBreakpoint = () => {
+  // Use efficient breakpoint detection that only triggers on actual breakpoint changes
+  const currentBreakpoint = useCurrentBreakpoint();
+
+  // Lazy up function - creates media query only when called
+  const up = useCallback((key: Breakpoint): boolean => {
+    const targetBreakpoint = breakpoints.find(bp => bp.id === key);
+    if (!targetBreakpoint) return false;
+
+    // For 'initial', always return true since we're always at or above 0px
+    if (key === 'initial') return true;
+
+    const query = `(min-width: ${targetBreakpoint.value}px)`;
+    return useSingleMediaQuery(query);
```

Comment:
These calls to `useSingleMediaQuery` aren't safe, both because the function is called `up` so it's not communicated to the caller that it's a hook, but also because of the conditions that can cause an early return.